pkg/util/retry.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+)
+
+type RetryConfig struct {
+	MaxAttempts     int
+	InitialInterval time.Duration
+	MaxInterval     time.Duration
+	Multiplier      float64
+	MaxElapsedTime  time.Duration
+}
+
+func DefaultRetryConfig() *RetryConfig {
+	return &RetryConfig{
+		MaxAttempts:     5,
+		InitialInterval: 1 * time.Second,
+		MaxInterval:     5 * time.Second,
+		Multiplier:      1.5,
+		MaxElapsedTime:  30 * time.Second,
+	}
+}
+
+type RetryableError interface {
+	ShouldRetry() bool
+}
+
+type RetryeError struct {
+	Err error
+}
+
+func (r RetryeError) Error() string {
+	return r.Err.Error()
+}
+
+func (r RetryeError) ShouldRetry() bool {
+	if r.Err == nil {
+		return false
+	}
+
+	// add all specific error messages that should trigger retry
+	retryableErrors := []string{
+		"error getting entries by index range",
```

Comment:
We shouldn't catch specific error messages, rather we should retry based on response codes.

The other part of this is figuring out how to handle http2 errors, specifically `http2: server sent GOAWAY and closed the connection`.